File: rxjava/src/main/java/com/asemlab/samples/rxjava/utils/NetworkUtils.java
package com.asemlab.samples.rxjava.utils;

import com.asemlab.samples.rxjava.remote.ApiResponseCallback;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class NetworkUtils {

    public static <T> void enqueueWithCallback(Call<T> call, ApiResponseCallback<T> apiCallback) {
        call.enqueue(new Callback<T>() {
            @Override
            public void onResponse(Call<T> call, Response<T> response) {
                if (response.isSuccessful() && response.body() != null) {
                    apiCallback.onSuccess(response.body());
                } else {
                    apiCallback.onError(response.code(), response.message());
                }
            }

            @Override
            public void onFailure(Call<T> call, Throwable t) {
                apiCallback.onFailure(t);
            }
        });
    }
}

File: inventory/src/main/java/com/asemlab/samples/inventory/utils/NetworkUtils.kt
package com.asemlab.samples.inventory.utils

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.util.Log
import com.asemlab.samples.inventory.remote.models.ServerResponse
import retrofit2.Response

suspend fun <T> performRequest(request: suspend () -> Response<T>): ServerResponse<T> {
    return try {
        val body = request()
        if (body.isSuccessful)
            ServerResponse.Success(body.body()!!)
        else
            ServerResponse.Error(
                body.errorBody()?.string() ?: "Empty error message",
                body.code()
            )
    } catch (e: Exception) {
        Log.e("performRequest", "${e.message}")
        throw e
    }
}


fun isConnected(context: Context): Boolean {
    val connectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    return connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
        ?.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED) == true
}